proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,950 @@
+lack of measurement traffic for some reason, the relay will always allow
+itself to send at least 10 background traffic cells each scheduling round.
+This worked fine for us in our real Internet experiments. Other strategies
+will be investigated that allow the relay to continue sending background
+traffic in the absence of measurement traffic while still trending towards
+some configured ratio.
+
+X.X FlashFlow code
+
+The FF coordinator and measurer code will reside in a FlashFlow repository
+separate from little-t tor.
+
+The coordinator is responsible for scheduling measurements, aggregating
+results, and producing v3bw files. It needs an accompanying Tor process so
+that it has access to consensuses as they come in.
+
+The measurers take commands from the coordinator, connect to target relays
+with many sockets, send them traffic, and verify the received traffic is the
+same as what was sent. Measurers either need to link against tor and call
+internal, undocumented, the-public-isnt-supposed-to-use-this-directly APIs
+(e.g. calling append_cell_to_circuit_queue() directly instead of using the
+limited but public interface exposed in feature/api/tor_api.c and
+tools/tor_runner.c), or tor needs some of this functionality added to it so
+that it can do some of these things itself when given commands via the control
+port.
+
+[XXX for now I'll assume that an optional little-t tor control port module
+housing a lot of this code is the best idea.]
+
+Notable new things that FlashFlow code in Tor will need to do:
+
+  X. Open many TLS+TCP connections to the same relay on purpose
+  X. Verify echo cells
+
+X.X.X Open many connections
+
+FlashFlow prototypes needed to "hack in" a flag in the
+open-a-connection-with-this-relay function call change that indicated whether
+or not we wanted to force a new connection to be created. Most of Tor doesn't
+care if it reuses an existing connection, but FF does want to create many
+different connections. The cleanest way to accomplish this will be
+investigated.
+
+X.X.X Verify echo cells
+
+A parameter will exist to tell the measurers with what frequency they shall
+verify cells echoed back to them match what they sent. This parameter does not
+to exist outside of the FF deployment (e.g. it doesn't need to be a consensus
+parameter).
+
+The parameter instructs the measurers to check 1 out of every N cells.
+
+The measurer logically splits its output stream of cells into buckets of size
+N. At the start of each bucket the measurer chooses a random index in the
+bucket. Upon sending the cell at that index, the measurer records it.
+The measurer also counts cells that it receives. When it receives a cell at an
+index that was recorded, it verifies that the received cell matches the
+recorded sent cell. If they match, no special action is taken. If they don't
+match, the measurer indicates failure to the coordinator and target relay and
+closes all connections, ending the measurement.
```

Comment:
This bucket thing is underspecified and confusing. I don't understand how to split an output stream into buckets, and I don't know what it means to be at the start of each bucket. That led me to be confused in the remainder of the security argument. Please explain; just adding a few more sentences should probably be enough.

Reply:
I've attempted to make it better and have included an example.